File: cpp/src/strategies.cpp
/// \file strategies.cpp
/// \brief Momentum, FX carry, and the walk-forward HMM regime gate.

#include "regime/strategies.hpp"

#include "regime/backtest.hpp"  // kTradingDays

#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>
#include <stdexcept>
#include <string>

namespace regime {

namespace {

constexpr double kNaN = std::numeric_limits<double>::quiet_NaN();

/// Validate a (T x A) return-like matrix: non-empty and all finite.
void validate_matrix(const Matrix& r, const char* name) {
    if (r.rows == 0 || r.cols == 0)
        throw std::invalid_argument(std::string(name) + " must be a non-empty 2-D array");
    for (double v : r.data)
        if (!std::isfinite(v)) throw std::invalid_argument(std::string(name) + " contain NaN or inf");
}

/// sign(x) in {-1, 0, +1}.
double sign(double x) { return x > 0.0 ? 1.0 : (x < 0.0 ? -1.0 : 0.0); }

}  // namespace

Matrix ewma_variance(const Matrix& returns, double lam, int init_window) {
    validate_matrix(returns, "returns");
    const std::size_t T = returns.rows, A = returns.cols;
    if (!(lam > 0.0 && lam < 1.0))
        throw std::invalid_argument("ewma lambda must be in (0, 1), got " + std::to_string(lam));
    if (init_window < 1 || static_cast<std::size_t>(init_window) > T)
        throw std::invalid_argument("init_window must be in [1, T], got " +
                                    std::to_string(init_window));
    const std::size_t w = static_cast<std::size_t>(init_window);
    Matrix sig2(T, A, kNaN);
    for (std::size_t a = 0; a < A; ++a) {
        double seed = 0.0;
        for (std::size_t t = 0; t < w; ++t) seed += returns(t, a) * returns(t, a);
        seed /= static_cast<double>(w);  // zero-mean convention
        sig2(w - 1, a) = seed;
        for (std::size_t t = w; t < T; ++t)
            sig2(t, a) = lam * sig2(t - 1, a) + (1.0 - lam) * returns(t, a) * returns(t, a);
    }
    return sig2;
}

Matrix momentum_positions(const Matrix& returns, int lookback, double vol_target,
                          double ewma_lambda, double leverage_cap) {
    validate_matrix(returns, "returns");
    const std::size_t T = returns.rows, A = returns.cols;
    if (lookback < 1)
        throw std::invalid_argument("lookback must be >= 1, got " + std::to_string(lookback));
    if (T <= static_cast<std::size_t>(lookback))
        throw std::invalid_argument("series shorter than momentum lookback: T=" +
                                    std::to_string(T) + " <= lookback=" + std::to_string(lookback));
    if (vol_target <= 0.0 || leverage_cap <= 0.0)
        throw std::invalid_argument("vol_target and leverage_cap must be > 0");
    const std::size_t lb = static_cast<std::size_t>(lookback);

    // Trailing cumulative return via cumulative growth ratios:
    // cumret[t] = growth[t] / growth[t - lookback] - 1 (growth[t] = prod(1+r)).
    Matrix growth(T, A);
    for (std::size_t a = 0; a < A; ++a) {
        double g = 1.0;
        for (std::size_t t = 0; t < T; ++t) {
            g *= 1.0 + returns(t, a);
            growth(t, a) = g;
        }
    }
    Matrix sig2 = ewma_variance(returns, ewma_lambda, lookback);

    Matrix pos(T, A, 0.0);
    for (std::size_t t = lb - 1; t < T; ++t) {
        for (std::size_t a = 0; a < A; ++a) {
            const double cumret = t == lb - 1 ? growth(t, a) - 1.0
                                              : growth(t, a) / growth(t - lb, a) - 1.0;
            const double ann_vol = std::sqrt(kTradingDays * sig2(t, a));
            const double lev = ann_vol > 0.0 ? std::min(vol_target / ann_vol, leverage_cap)
                                             : leverage_cap;
            pos(t, a) = sign(cumret) * lev / static_cast<double>(A);
        }
    }
    return pos;
}

Matrix carry_positions(const Matrix& rate_diffs, int top_n, int bottom_n,
                       int rebalance_days) {
    validate_matrix(rate_diffs, "rate differentials");
    const std::size_t T = rate_diffs.rows, A = rate_diffs.cols;
    if (top_n < 1 || bottom_n < 1 ||
        static_cast<std::size_t>(top_n) + static_cast<std::size_t>(bottom_n) > A)
        throw std::invalid_argument("need top_n>=1, bottom_n>=1, top_n+bottom_n<=A=" +
                                    std::to_string(A) + "; got " + std::to_string(top_n) + "," +
                                    std::to_string(bottom_n));
    if (rebalance_days < 1)
        throw std::invalid_argument("rebalance_days must be >= 1, got " +
                                    std::to_string(rebalance_days));

    Matrix pos(T, A, 0.0);
    std::vector<double> current(A, 0.0);
    std::vector<std::size_t> order(A);
    for (std::size_t t = 0; t < T; ++t) {
        if (t % static_cast<std::size_t>(rebalance_days) == 0) {
            // Stable sort on descending differential; asset index breaks ties.
            std::iota(order.begin(), order.end(), 0);
            std::stable_sort(order.begin(), order.end(),
                             [&rate_diffs, t](std::size_t a, std::size_t b) {
                                 return rate_diffs(t, a) > rate_diffs(t, b);
                             });
            std::fill(current.begin(), current.end(), 0.0);
            for (int i = 0; i < top_n; ++i)
                current[order[static_cast<std::size_t>(i)]] = 1.0 / top_n;
            for (int i = 0; i < bottom_n; ++i)
                current[order[A - 1 - static_cast<std::size_t>(i)]] = -1.0 / bottom_n;
        }
        for (std::size_t a = 0; a < A; ++a) pos(t, a) = current[a];
    }
    return pos;
}

Matrix carry_total_returns(const Matrix& spot_returns, const Matrix& rate_diffs) {
    validate_matrix(spot_returns, "spot returns");
    validate_matrix(rate_diffs, "rate differentials");
    if (spot_returns.rows != rate_diffs.rows || spot_returns.cols != rate_diffs.cols)
        throw std::invalid_argument("spot returns and differentials must have equal shape");
    Matrix total = spot_returns;
    for (std::size_t t = 1; t < total.rows; ++t)
        for (std::size_t a = 0; a < total.cols; ++a)
            total(t, a) += rate_diffs(t - 1, a) / kTradingDays;
    return total;
}

RegimeGateResult regime_gate(const std::vector<double>& index_returns, int n_states,
                             int train_min_days, int refit_days, const std::string& mode,
                             double threshold, double tol, int max_iter, double var_floor) {
    for (double v : index_returns)
        if (!std::isfinite(v)) throw std::invalid_argument("index_returns contain NaN or inf");
    if (mode != "prob" && mode != "binary")
        throw std::invalid_argument("gate mode must be 'prob' or 'binary', got '" + mode + "'");
    const std::size_t T = index_returns.size();
    if (train_min_days <= n_states || refit_days < 1)
        throw std::invalid_argument(
            "train_min_days must exceed n_states and refit_days must be >= 1");
    if (T < static_cast<std::size_t>(train_min_days))
        throw std::invalid_argument("series shorter than train_min_days: T=" + std::to_string(T) +
                                    " < " + std::to_string(train_min_days));

    const std::size_t t0 = static_cast<std::size_t>(train_min_days) - 1;
    RegimeGateResult out;
    out.gate.assign(T, 1.0);
    std::vector<double> alpha;
    std::size_t calm_state = 0;
    for (std::size_t t = t0; t < T; ++t) {
        if ((t - t0) % static_cast<std::size_t>(refit_days) == 0) {
            GaussianHMM model(n_states, tol, max_iter, var_floor);
            Matrix window(t + 1, 1);
            for (std::size_t u = 0; u <= t; ++u) window(u, 0) = index_returns[u];
            if (out.models.empty()) {
                model.fit(window);
            } else {
                model.fit(window, out.models.back().params());  // warm start
            }
            // Pinned gate state: lowest variance on dimension 0 (ties -> lower label).
            const Matrix& vars = model.params().variances;
            calm_state = 0;
            for (std::size_t k = 1; k < vars.rows; ++k)
                if (vars(k, 0) < vars(calm_state, 0)) calm_state = k;
            Matrix filt = model.filtered_probabilities(window);
            alpha.assign(filt.cols, 0.0);
            for (std::size_t k = 0; k < filt.cols; ++k) alpha[k] = filt(t, k);
            out.models.push_back(std::move(model));
        } else {
            alpha = forward_step(out.models.back().params(), alpha, {index_returns[t]});
        }
        const double p_calm = alpha[calm_state];
        out.gate[t] = mode == "prob" ? p_calm : (p_calm >= threshold ? 1.0 : 0.0);
    }
    return out;
}

Matrix apply_gate(const Matrix& positions, const std::vector<double>& gate) {
    validate_matrix(positions, "positions");
    if (gate.size() != positions.rows)
        throw std::invalid_argument("gate length does not match positions");
    for (double g : gate)
        if (!std::isfinite(g)) throw std::invalid_argument("gate contains NaN or inf");
    Matrix out = positions;
    for (std::size_t t = 0; t < out.rows; ++t)
        for (std::size_t a = 0; a < out.cols; ++a) out(t, a) *= gate[t];
    return out;
}

}  // namespace regime

File: rust/src/strategies.rs
//! Trading strategies: time-series momentum, FX carry, and the HMM regime gate.
//!
//! # No-lookahead convention (shared with the backtest engine)
//!
//! Every position matrix `P` produced here has `P[t]` = the position
//! *decided at the close of day t using information up to and including day
//! t*.  The backtest applies `P[t-1]` to the day-t return, so a signal can
//! never see the return it is traded against.  The regime gate at day t
//! likewise uses only observations `x_0..x_t` (filtered, not smoothed,
//! probabilities) and models fitted on data through day t.

use crate::error::{RegimeError, Result};
use crate::hmm::{forward_step, GaussianHmm};
use crate::matrix::Matrix;

/// Trading days per year (pinned).
pub const TRADING_DAYS: f64 = 252.0;

fn validate_returns(r: &Matrix, name: &str) -> Result<()> {
    if r.rows() == 0 || r.cols() == 0 {
        return Err(RegimeError::InvalidInput(format!("{name} must be a non-empty matrix")));
    }
    if !r.all_finite() {
        return Err(RegimeError::InvalidInput(format!("{name} contain NaN or inf")));
    }
    Ok(())
}

/// EWMA variance per asset with a pinned initialization.
///
/// At `t = init_window - 1` the variance is the mean of squared returns over
/// the first `init_window` days (zero-mean convention); for
/// `t >= init_window`:
///
/// ```text
/// sigma2[t] = lam * sigma2[t-1] + (1 - lam) * r[t]^2
/// ```
///
/// Entries before `init_window - 1` are NaN (undefined).
///
/// # Errors
/// Invalid returns, `lam` outside `(0, 1)`, or `init_window` outside `[1, T]`.
pub fn ewma_variance(returns: &Matrix, lam: f64, init_window: usize) -> Result<Matrix> {
    validate_returns(returns, "returns")?;
    let (t_len, n_assets) = (returns.rows(), returns.cols());
    if !(lam > 0.0 && lam < 1.0) {
        return Err(RegimeError::InvalidInput(format!("ewma lambda must be in (0, 1), got {lam}")));
    }
    if init_window < 1 || init_window > t_len {
        return Err(RegimeError::InvalidInput(format!(
            "init_window must be in [1, T], got {init_window}"
        )));
    }
    let mut sig2 = Matrix::filled(t_len, n_assets, f64::NAN);
    for a in 0..n_assets {
        let mut acc = 0.0;
        for t in 0..init_window {
            let v = returns.get(t, a);
            acc += v * v;
        }
        sig2.set(init_window - 1, a, acc / init_window as f64);
        for t in init_window..t_len {
            let r = returns.get(t, a);
            let prev = sig2.get(t - 1, a);
            sig2.set(t, a, lam * prev + (1.0 - lam) * r * r);
        }
    }
    Ok(sig2)
}

/// Time-series momentum positions with vol-targeted sizing (API_SPEC.md §3.1).
///
/// The signal at day t (defined for `t >= lookback - 1`) is the sign of the
/// trailing cumulative simple return `prod(1 + r) - 1` over the `lookback`
/// days ending at t.  Sizing: per-asset leverage
/// `min(vol_target / ann_vol[t], leverage_cap)` with
/// `ann_vol = sqrt(252 * sigma2_EWMA)`; zero EWMA vol gets the cap.
/// Positions are divided by the asset count and are zero before the first
/// full lookback window.
///
/// # Errors
/// `T <= lookback`, non-positive `vol_target`/`leverage_cap`, or invalid
/// returns/lambda.
pub fn momentum_positions(
    returns: &Matrix,
    lookback: usize,
    vol_target: f64,
    ewma_lambda: f64,
    leverage_cap: f64,
) -> Result<Matrix> {
    validate_returns(returns, "returns")?;
    let (t_len, n_assets) = (returns.rows(), returns.cols());
    if lookback < 1 {
        return Err(RegimeError::InvalidInput(format!("lookback must be >= 1, got {lookback}")));
    }
    if t_len <= lookback {
        return Err(RegimeError::InvalidInput(format!(
            "series shorter than momentum lookback: T={t_len} <= lookback={lookback}"
        )));
    }
    if vol_target <= 0.0 || leverage_cap <= 0.0 {
        return Err(RegimeError::InvalidInput("vol_target and leverage_cap must be > 0".into()));
    }

    // Trailing cumulative return via cumulative growth ratios (matches the
    // reference: growth[t] / growth[t - lookback] - 1).
    let mut growth = Matrix::zeros(t_len, n_assets);
    for a in 0..n_assets {
        let mut acc = 1.0;
        for t in 0..t_len {
            acc *= 1.0 + returns.get(t, a);
            growth.set(t, a, acc);
        }
    }
    let sig2 = ewma_variance(returns, ewma_lambda, lookback)?;

    let mut pos = Matrix::zeros(t_len, n_assets);
    for t in (lookback - 1)..t_len {
        for a in 0..n_assets {
            let cumret = if t == lookback - 1 {
                growth.get(t, a) - 1.0
            } else {
                growth.get(t, a) / growth.get(t - lookback, a) - 1.0
            };
            let signal = if cumret > 0.0 {
                1.0
            } else if cumret < 0.0 {
                -1.0
            } else {
                0.0
            };
            let ann_vol = (TRADING_DAYS * sig2.get(t, a)).sqrt();
            let lev = if ann_vol > 0.0 {
                (vol_target / ann_vol).min(leverage_cap)
            } else {
                leverage_cap
            };
            pos.set(t, a, signal * lev / n_assets as f64);
        }
    }
    Ok(pos)
}

/// FX carry positions: long high-differential, short low-differential (§3.2).
///
/// Pinned schedule: rebalance at day indices `0, rebalance_days, ...`;
/// positions held unchanged in between.  At each rebalance, currencies are
/// ranked by that day's differential, descending, with ties broken by
/// ascending asset index (stable sort).  The top `top_n` get `+1/top_n`
/// each, the bottom `bottom_n` get `-1/bottom_n` each, the rest 0.
///
/// # Errors
/// `top_n + bottom_n > A`, zero `top_n`/`bottom_n`/`rebalance_days`, or
/// invalid differentials.
pub fn carry_positions(
    rate_diffs: &Matrix,
    top_n: usize,
    bottom_n: usize,
    rebalance_days: usize,
) -> Result<Matrix> {
    validate_returns(rate_diffs, "rate differentials")?;
    let (t_len, n_assets) = (rate_diffs.rows(), rate_diffs.cols());
    if top_n < 1 || bottom_n < 1 || top_n + bottom_n > n_assets {
        return Err(RegimeError::InvalidInput(format!(
            "need top_n>=1, bottom_n>=1, top_n+bottom_n<=A={n_assets}; got {top_n},{bottom_n}"
        )));
    }
    if rebalance_days < 1 {
        return Err(RegimeError::InvalidInput(format!(
            "rebalance_days must be >= 1, got {rebalance_days}"
        )));
    }

    let mut pos = Matrix::zeros(t_len, n_assets);
    let mut current = vec![0.0; n_assets];
    for t in 0..t_len {
        if t % rebalance_days == 0 {
            // Stable sort, descending differential; index breaks ties.
            let mut order: Vec<usize> = (0..n_assets).collect();
            order.sort_by(|&a, &b| {
                rate_diffs
                    .get(t, b)
                    .partial_cmp(&rate_diffs.get(t, a))
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
            current = vec![0.0; n_assets];
            for &i in order.iter().take(top_n) {
                current[i] = 1.0 / top_n as f64;
            }
            for &i in order.iter().skip(n_assets - bottom_n) {
                current[i] = -1.0 / bottom_n as f64;
            }
        }
        for a in 0..n_assets {
            pos.set(t, a, current[a]);
        }
    }
    Ok(pos)
}

/// Total currency returns: spot move plus accrued carry (§3.2).
///
/// Pinned accrual: the day-t total return is `spot[t] + diff[t-1] / 252` —
/// the differential set at the previous close (known in advance, no
/// lookahead).  Day 0 accrues nothing.
///
/// # Errors
/// Shape mismatch or invalid input.
pub fn carry_total_returns(spot_returns: &Matrix, rate_diffs: &Matrix) -> Result<Matrix> {
    validate_returns(spot_returns, "spot returns")?;
    validate_returns(rate_diffs, "rate differentials")?;
    if spot_returns.rows() != rate_diffs.rows() || spot_returns.cols() != rate_diffs.cols() {
        return Err(RegimeError::InvalidInput(
            "spot returns and differentials must have equal shape".into(),
        ));
    }
    let mut total = spot_returns.clone();
    for t in 1..total.rows() {
        for a in 0..total.cols() {
            let v = total.get(t, a) + rate_diffs.get(t - 1, a) / TRADING_DAYS;
            total.set(t, a, v);
        }
    }
    Ok(total)
}

/// Walk-forward HMM regime gate on a market proxy (API_SPEC.md §3.3).
///
/// Pinned schedule: the first fit happens at day `t0 = train_min_days - 1`
/// on the expanding window `r[0..=t0]`; refits at `t0 + refit_days * k` on
/// `r[0..=t]`.  The first fit starts from the pinned quantile
/// initialization; every later refit warm-starts from the previous fitted
/// (label-sorted) parameters.
///
/// The gate at day t (for `t >= t0`) is the *filtered* probability
/// `p_calm(t) = P(s_t = k* | r_0..r_t)` of the low-vol state under the most
/// recent model, where `k* = argmin_k variances[k, 0]` (ties -> lower
/// label).  A full scaled forward pass runs at each refit day; single
/// [`forward_step`] updates are used in between (identical math, no
/// lookahead).  Before the first fit the gate is 1.
///
/// `mode = "prob"` returns `gate[t] = p_calm(t)`; `mode = "binary"` returns
/// `1.0` if `p_calm(t) >= threshold` else `0.0`.
///
/// Returns the gate values in `[0, 1]` and the fitted models in refit order.
///
/// # Errors
/// Series shorter than `train_min_days`, `train_min_days <= n_states`,
/// `refit_days < 1`, unknown `mode`, or non-finite input.
#[allow(clippy::too_many_arguments)]
pub fn regime_gate(
    index_returns: &[f64],
    n_states: usize,
    train_min_days: usize,
    refit_days: usize,
    mode: &str,
    threshold: f64,
    tol: f64,
    max_iter: usize,
    var_floor: f64,
) -> Result<(Vec<f64>, Vec<GaussianHmm>)> {
    if index_returns.iter().any(|v| !v.is_finite()) {
        return Err(RegimeError::InvalidInput("index_returns contain NaN or inf".into()));
    }
    if mode != "prob" && mode != "binary" {
        return Err(RegimeError::InvalidInput(format!(
            "gate mode must be 'prob' or 'binary', got {mode:?}"
        )));
    }
    let t_total = index_returns.len();
    if train_min_days <= n_states || refit_days < 1 {
        return Err(RegimeError::InvalidInput(
            "train_min_days must exceed n_states and refit_days must be >= 1".into(),
        ));
    }
    if t_total < train_min_days {
        return Err(RegimeError::InvalidInput(format!(
            "series shorter than train_min_days: T={t_total} < {train_min_days}"
        )));
    }

    let full = Matrix::column(index_returns);
    let t0 = train_min_days - 1;
    let mut gate = vec![1.0; t_total];
    let mut models: Vec<GaussianHmm> = Vec::new();
    let mut alpha: Vec<f64> = Vec::new();
    let mut calm_state = 0usize;
    for t in t0..t_total {
        if (t - t0) % refit_days == 0 {
            let mut model = GaussianHmm::new(n_states, tol, max_iter, var_floor)?;
            let warm = models.last().and_then(|m| m.params()).cloned();
            let window = full.head(t + 1);
            model.fit(&window, warm.as_ref())?;
            let params = model.params().expect("fit just succeeded");
            calm_state = argmin_variance(params);
            let filtered = model.filtered_probabilities(&window)?;
            alpha = filtered.row(t).to_vec();
            models.push(model);
        } else {
            let params = models.last().and_then(|m| m.params()).expect("model fitted");
            alpha = forward_step(params, &alpha, &[index_returns[t]]);
        }
        let p_calm = alpha[calm_state];
        gate[t] = if mode == "prob" {
            p_calm
        } else if p_calm >= threshold {
            1.0
        } else {
            0.0
        };
    }
    Ok((gate, models))
}

/// Index of the lowest-variance state on dimension 0 (ties -> lower label).
fn argmin_variance(params: &crate::hmm::HmmParams) -> usize {
    let k = params.startprob.len();
    let mut best = 0;
    for i in 1..k {
        if params.variances.get(i, 0) < params.variances.get(best, 0) {
            best = i;
        }
    }
    best
}

/// Scale each day's positions by that day's gate value.
///
/// `P_filtered[t] = gate[t] * P[t]`; both are day-t decisions, so the
/// filtered position still only trades against the day t+1 return.
///
/// # Errors
/// Length mismatch or non-finite gate values.
pub fn apply_gate(positions: &Matrix, gate: &[f64]) -> Result<Matrix> {
    validate_returns(positions, "positions")?;
    if gate.len() != positions.rows() {
        return Err(RegimeError::InvalidInput(format!(
            "gate length {} does not match positions rows {}",
            gate.len(),
            positions.rows()
        )));
    }
    if gate.iter().any(|v| !v.is_finite()) {
        return Err(RegimeError::InvalidInput("gate contains NaN or inf".into()));
    }
    let mut out = positions.clone();
    for t in 0..out.rows() {
        for a in 0..out.cols() {
            out.set(t, a, positions.get(t, a) * gate[t]);
        }
    }
    Ok(out)
}
